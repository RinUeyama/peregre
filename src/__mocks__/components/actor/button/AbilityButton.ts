import { IAbilityButtonProps } from 'src/components/actor/button/AbilityButton';
import { INITIAL_ABILITY } from 'src/constants';

const MOCK_ABILITY_BUTTON_PROPS: IAbilityButtonProps = { ability: INITIAL_ABILITY };

export default MOCK_ABILITY_BUTTON_PROPS;
