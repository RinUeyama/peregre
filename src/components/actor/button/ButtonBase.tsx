
import React from 'react'
import styled, { css } from 'styled-components'
import { color, space } from '~/assets/style'

export interface IButtonBaseProps {
  children?: React.ReactNode
  isEditMode?: boolean
  type?: 'button' | 'submit' | 'reset'
  onClick?: React.MouseEventHandler<HTMLButtonElement>
  unset?: boolean
  lighten?: boolean
  name?: string;
  value?: any;
}

export const ButtonBase: React.FC<IButtonBaseProps> = ({ children, isEditMode = false, type = 'button', onClick, unset = false, lighten = false, name, value }) => {
  return (
    <ButtonBaseWrapper type={type} onClick={onClick} isEditMode={isEditMode} unset={unset} lighten={lighten} name={name} value={value}>
      {children}
    </ButtonBaseWrapper>
  )
}

// TODO: 押した時に中身を若干小さくする
// TODO: もっとかわいい感じのアニメーション
// TODO: 型推論が効いていない
const ButtonBaseWrapper = styled.button<IButtonBaseProps>`
  width: 100%;
  height: 100%;
  display: flex;
  flex-direction: column;
  justify-content: center;
  align-items: center;
  background-color: ${color.backgroundHighContrast};
  color: ${color.fontInHighContrast};
  padding: ${space.xs};
  border: 0;
  border-radius: inherit;
  outline: 0;
  transition: box-shadow ease-out 0.3s;
  font-size: calc(0.8rem + 0.5vmin);
  line-height: 1.8;
  box-shadow:  2px 2px 6px ${color.backgroundLowContrastShadowDarken},
                 -2px -2px 6px ${color.backgroundLowContrastShadowLighten};
  &:active {
    box-shadow: inset 2px 2px 6px ${color.backgroundLowContrastShadowDarken},
                inset -2px -2px 6px ${color.backgroundLowContrastShadowLighten};
  }

  ${props => props.isEditMode && css `
    border-radius: 20px;
  `}

  ${props => props.unset && css `
    color: ${color.background};
    background-color: ${color.font};
  `}

  ${props => props.lighten && css`
    color: ${color.font};
    background-color: ${color.background};
  `}
`
