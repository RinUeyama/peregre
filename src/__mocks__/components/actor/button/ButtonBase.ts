import { IButtonBaseProps } from '~/components/actor/button/ButtonBase';

const MOCK_BUTTON_BASE_PROPS: IButtonBaseProps = {};

export default MOCK_BUTTON_BASE_PROPS;
