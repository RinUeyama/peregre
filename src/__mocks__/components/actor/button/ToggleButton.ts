import { IToggleButtonProps } from 'src/components/actor/button/ToggleButton'
import { Icons } from 'src/assets/icons'

const MOCK_TOGGLE_BUTTON_PROPS: IToggleButtonProps = {
  defaultImage: {
    src: Icons.Hidden,
    alt: 'でふぉ',
  },
  defaultLabel: 'ON',
  reversedImage: {
    src: Icons.Enlightenment,
    alt: 'ひっくり返ってる',
  },
  reversedLabel: 'OFF',
}

export default MOCK_TOGGLE_BUTTON_PROPS
