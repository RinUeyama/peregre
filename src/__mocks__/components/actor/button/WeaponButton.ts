import { IWeaponButtonProps } from 'src/components/actor/button/WeaponButton'
import { INITIAL_SKILL } from 'src/constants'

const MOCK_WEAPON_BUTTON_PROPS: IWeaponButtonProps = {
  skill: INITIAL_SKILL,
  setCurrentSkill: console.log('setCurrentSkill run'),
}

export default MOCK_WEAPON_BUTTON_PROPS
