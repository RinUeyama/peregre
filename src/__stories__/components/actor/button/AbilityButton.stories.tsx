import React from 'react'
import { Story, Meta } from '@storybook/react'
import { AbilityButton, IAbilityButtonProps } from '~/components/actor/button/AbilityButton'
import MOCK_ABILITY_BUTTON_PROPS from '~/__mocks__/components/actor/button/AbilityButton'

export default {
  title: 'components/actor/button/AbilityButton',
  component: AbilityButton,
  decorators: [(Story) => <div style={{ padding: '32px', backgroundColor: '#fef8e7' }}><Story/></div>]
} as Meta

const Template: Story<IAbilityButtonProps> = (args) => <AbilityButton {...args} />
export const Default = Template.bind({})
Default.args = {
  ...MOCK_ABILITY_BUTTON_PROPS
}