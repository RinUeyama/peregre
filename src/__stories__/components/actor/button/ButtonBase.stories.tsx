import React from 'react'
import { Story, Meta } from '@storybook/react'
import { ButtonBase, IButtonBaseProps } from '~/components/actor/button/ButtonBase'
import MOCK_BUTTON_BASE_PROPS from '~/__mocks__/components/actor/button/ButtonBase'

export default {
  title: 'Components/actor/button/ButtonBase',
  decorators: [(Story) => <div style={{ padding: '32px', backgroundColor: '#fef8e7' }}><Story/></div>]
} as Meta

const Template: Story<IButtonBaseProps> = (props) => <ButtonBase {...props} />
export const Default = Template.bind({})
Default.args = {
  ...MOCK_BUTTON_BASE_PROPS
}