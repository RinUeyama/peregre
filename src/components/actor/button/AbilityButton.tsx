import React from 'react'
import styled from 'styled-components'
import { space } from 'src/assets/style'
import { IRollResult, ROLL_RESULT } from 'src/constants'
import { IAbility } from 'src/interfaces'
import { rollDice10 } from 'src/utils/Math'
import { ButtonBase } from './ButtonBase'

export interface IAbilityButtonProps {
  ability: IAbility;
}

// dummy function
const sendMessageToDiscord = (type: string, message: string) => {
  console.log(type)
  console.log(message)
  return
}

export const AbilityButton: React.VFC<IAbilityButtonProps> = ({ ability }) => {
  const handleClick = () => {
    if (ability == null) return

    const diceNumber = rollDice10()
    const rollResult: IRollResult = (() => {
      if (diceNumber === 1) return ROLL_RESULT.CRITICAL
      else if (diceNumber === 10) return ROLL_RESULT.FUMBLE
      else if (diceNumber < 3 * ability.level) return ROLL_RESULT.SUCCESS
      else return ROLL_RESULT.FAILURE
    })()
    const message = `${ability.name}: ${diceNumber}${rollResult}`
    sendMessageToDiscord('ability', message)
  }

  return (
    <AbilityButtonWrapper>
      {/** TODO: 文言組み込み */}
      <ButtonBase onClick={handleClick} unset={ability.name === '未設定'}>
        {/** TODO: NO_IMAGE */}
        <Icon {...ability.icon} />
        <Label>{ability.name}</Label>
      </ButtonBase>
    </AbilityButtonWrapper>
  )
}

const AbilityButtonWrapper = styled.div`
  width: 120px;
  height: 120px;
`

const Icon = styled.img`
  padding: ${space.xxs};
`

const Label = styled.p`
`