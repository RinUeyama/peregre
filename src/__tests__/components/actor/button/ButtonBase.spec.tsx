import { ButtonBase } from 'src/components/actor/button/ButtonBase';
import MOCK_BUTTON_BASE_PROPS from 'src/__mocks__/components/actor/button/ButtonBase';
import { render } from '@testing-library/react';

describe('actor/button', () => {
  const snapshot = render(<ButtonBase {...MOCK_BUTTON_BASE_PROPS}/>);

  it('default test', () => {
    expect(snapshot.baseElement).toMatchSnapshot();
  });
});
