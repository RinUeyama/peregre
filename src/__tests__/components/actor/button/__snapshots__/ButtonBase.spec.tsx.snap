// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`actor/button default test 1`] = `
<body>
  <div>
    <button
      class="sc-bdvvaa jExhKM"
      type="button"
    />
  </div>
</body>
`;
